parseusermultihypo assertion error

Hi, 

The assertion `@assert sum(multihypo2) % 1 ≈ 0 `  in `parseusermultihypo` throws an error if the user input is slightly below some integer:

```
ERROR: LoadError: AssertionError: sum(multihypo2) % 1 ≈ 0[ Info: monitorCSMs: all tasks done

Stacktrace:
 [1] parseusermultihypo(::Array{Float64,1}, ::Float64) at /home/leopold/.julia/packages/IncrementalInference/JEqfk/src/FactorGraph.jl:544
```

Which can be easily reproduced on the repl:

```julia-repl
julia> failMH = measurements["lm"]["jLm"]["p"][12]
2×1 Array{Float64,2}:
 0.4958648003278645
 0.5041351996721354

julia> res = sum(failMH) % 1
0.9999999999999999

julia> res ≈ 0
false
```

Modulo of some value slightly below some integer will always be 0.999... and not 0.

Perhaps a check function like 

```julia
function isapproxinteger(x)
	return x ≈ round(x)
end
```

used in an assertion like 

```julia
@assert isapproxinteger(sum(multihypo2))
```

would solve the problem, in the meantime i can always correct my probability input using 

```julia
p ./= sum(p)
```
but a more numerical-error-tolerant assertion would be good anyways.

P.S.: Using `≈` in against 0 is like using `==` (https://docs.julialang.org/en/v1/base/math/#Base.isapprox):
Note that x ≈ 0 (i.e., comparing to zero with the default tolerances) is equivalent to x == 0 since the default atol is 0. In such cases, you should either supply an appropriate atol (or use norm(x) ≤ atol) or rearrange your code (e.g. use x ≈ y rather than x - y ≈ 0). It is not possible to pick a nonzero atol automatically because it depends on the overall scaling (the "units") of your problem: for example, in x - y ≈ 0, atol=1e-9 is an absurdly small tolerance if x is the radius of the Earth in meters, but an absurdly large tolerance if x is the radius of a Hydrogen atom in meters.

--- a/Project.toml
+++ b/Project.toml
@@ -2,7 +2,7 @@ name = "IncrementalInference"
 uuid = "904591bb-b899-562f-9e6f-b8df64c7d480"
 keywords = ["mm-iSAM", "SLAM", "inference", "sum-product", "belief-propagation"]
 desc = "Implements the multimodal iSAM algorithm."
-version = "0.18.1"
+version = "0.18.2"
 
 [deps]
 ApproxManifoldProducts = "9bbbb610-88a1-53cd-9763-118ce10c1f89"

--- a/src/FactorGraph.jl
+++ b/src/FactorGraph.jl
@@ -541,16 +541,12 @@ function parseusermultihypo(multihypo::Vector{Float64}, nullhypo::Float64)
     multihypo2 = multihypo
     multihypo2[1-1e-10 .< multihypo] .= 0.0
     # check that terms sum to full probability
-    @assert sum(multihypo2) % 1 ≈ 0
+    @assert sum(multihypo2) % 1 ≈ 0.0 || 1-1e-10 < sum(multihypo2) % 1 "ensure multihypo sums to a (or nearly, 1e-10) interger, see #1086"
     # check that only one variable broken into fractions
     @assert sum(multihypo2[1e-10 .< multihypo2]) ≈ 1
-    # multihypo2 = Float64[multihypo...]
-    # verts = Symbol.(multihypo[1,:])
-    # for i in 1:length(multihypo)
-    #   if multihypo[i] > 0.999999
-    #     multihypo2[i] = 0.0
-    #   end
-    # end
+    # force normalize something that is now known to be close
+    multihypo2 ./= sum(multihypo2)
+
     mh = Categorical(Float64[multihypo2...] )
   end
   return mh, nullhypo

--- a/test/testmultihypothesisapi.jl
+++ b/test/testmultihypothesisapi.jl
@@ -9,6 +9,7 @@ using Statistics
 import Base: convert
 import IncrementalInference: getSample
 
+##
 
 mutable struct DevelopPrior <: AbstractPrior
   x::Distribution
@@ -250,6 +251,20 @@ global pts = approxConv(fg, :x2x3x4x5f1, :x5, N=N)
 
 @test 0.5*N <= sum(80 .< pts .< 100.0) + sum(pts .== 2.0) + sum(pts .== 3.0)
 
+
+
+end
+
+##
+
+@testset "test multihypo api numerical tolerance, #1086" begin
+
+fg = initfg()
+
+addVariable!(fg, :x0, ContinuousEuclid{1})
+addVariable!(fg, :x1a, ContinuousEuclid{1})
+addVariable!(fg, :x1b, ContinuousEuclid{1})
+addFactor!(fg, [:x0;:x1a;:x1b], LinearRelative(Normal()), multihypo=[1; 0.5;0.4999999999999])
 
 
 end

